sidekick_core/lib/src/sidekick_version_checker.dart

```diff
@@ -0,0 +1,122 @@
+import 'dart:convert';
+
+import 'package:http/http.dart';
+import 'package:pub_semver/pub_semver.dart';
+import 'package:sidekick_core/sidekick_core.dart';
+import 'package:yaml/yaml.dart';
+
+/// Checks and updates dependencies of the generated sidekick CLI
+class SidekickVersionChecker {
+  const SidekickVersionChecker();
+
+  /// Checks whether the latest version of [package] is being used in the generated sidekick CLI
+  ///
+  /// [pubspecPath] can be passed to override the default behavior of reading
+  /// the current package version from the pubspec.yaml at ['dependencies'][package]
+  /// E.g. ['dev_dependencies', 'some_dev_dependency'] or ['sidekick', 'cli_version']
+  Future<bool> isUpToDate({
+    required String package,
+    List<String>? pubspecPath,
+  }) async {
+    final latest = await getLatestPackageVersion(package);
+    final current = getCurrentMinimumPackageVersion(
+      pubspecPath ?? ['dependencies', package],
+    );
+
+    return latest == current;
+  }
+
+  /// Updates the version constraint of [package] to the latest version
+  Future<void> updateVersionConstraintToLatest(String package) async =>
+      updateVersionConstraint(
+        package: package,
+        newMinimumVersion: await getLatestPackageVersion(package),
+      );
+
+  /// Replaces the current version constraint of [package] in pubspec.yaml with [newMinimumVersion]
+  ///
+  /// If [pinVersion] is true, the new version constraint will only allow [newMinimumVersion]
+  /// Else it will allow a version range from [newMinimumVersion] until the next breaking version
+  void updateVersionConstraint({
+    required String package,
+    required Version newMinimumVersion,
+    bool pinVersion = false,
+  }) {
+    final pubspec = Repository.requiredSidekickPackage.pubspec;
+    final lines = pubspec.readAsLinesSync();
+
+    final newVersionConstraint = pinVersion
+        ? newMinimumVersion.canonicalizedVersion
+        : newMinimumVersion.major > 0
+            ? '^${newMinimumVersion.canonicalizedVersion}'
+            : "'>=${newMinimumVersion.canonicalizedVersion} <1.0.0'";
+
+    final index = lines.indexWhere((it) => it.startsWith('  $package:'));
+    lines[index] = '  $package: $newVersionConstraint';
```

Comment:
```
Got dependencies!
Sidekick Warning: No Dart SDK set. Please set it in `initializeSidekick(dartSdkPath: 'path/to/sdk')`
Updating sidekick CLI aloh from version 0.0.0 to 0.11.1 ...
There was an error updating sidekick CLI aloh from version 0.0.0 to 0.11.1.
Unhandled exception:
RangeError (index): Invalid value: Not in inclusive range 0..23: -1
#0      List._setIndexed (dart:core-patch/growable_array.dart:273:49)
#1      List.[]= (dart:core-patch/growable_array.dart:268:5)
#2      SidekickVersionChecker.updateVersionConstraint (package:sidekick_core/src/sidekick_version_checker.dart:55:10)
#3      main (package:sidekick_core/src/update_sidekick_cli.dart:53:28)
#4      main (file:///Users/pascalwelsch/Projects/phntmxyz/Alohomora/aloh_sidekick/build/update.dart:3:16)
#5      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:32)
#6      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)
Sidekick Warning: No Dart SDK set. Please set it in `initializeSidekick(dartSdkPath: 'path/to/sdk')`
```

pubspec:

```
dependencies:
  csv: ^5.0.0
  dcli: any
  intl: ^0.17.0
  sidekick_core: '>=0.11.1 <1.0.0'
  timeago: ^3.0.2
```

Reply:
This should be fixed now, I had a lot of fun writing the regex for my solution :D I've added some tests for these edge cases as well